src/Databases/DDLDependencyVisitor.cpp

```diff
@@ -1,198 +1,239 @@
-
-TableNamesSet getDependenciesSetFromCreateQuery(ContextPtr global_context, const QualifiedTableName & table, const ASTPtr & ast)
-{
-    assert(global_context == global_context->getGlobalContext());
-    TableLoadingDependenciesVisitor::Data data;
-    data.default_database = global_context->getCurrentDatabase();
-    data.create_query = ast;
-    data.global_context = global_context;
-    TableLoadingDependenciesVisitor visitor{data};
-    visitor.visit(ast);
-    data.dependencies.erase(table);
-    return data.dependencies;
-}
-
-void DDLDependencyVisitor::visit(const ASTPtr & ast, Data & data)
+namespace DB
 {
-    /// Looking for functions in column default expressions and dictionary source definition
-    if (const auto * function = ast->as<ASTFunction>())
-        visit(*function, data);
-    else if (const auto * dict_source = ast->as<ASTFunctionWithKeyValueArguments>())
-        visit(*dict_source, data);
-    else if (const auto * storage = ast->as<ASTStorage>())
-        visit(*storage, data);
-}
 
-bool DDLMatcherBase::needChildVisit(const ASTPtr & node, const ASTPtr & child)
+namespace
 {
-    if (node->as<ASTStorage>())
-        return false;
-
-    if (auto * create = node->as<ASTCreateQuery>())
+    /// CREATE TABLE or CREATE DICTIONARY or CREATE VIEW or CREATE TEMPORARY TABLE or CREATE DATABASE query.
+    void visitCreateQuery(const ASTCreateQuery & create, DDLDependencyVisitor::Data & data)
     {
-        if (child.get() == create->select)
-            return false;
-    }
+        QualifiedTableName to_table{create.to_table_id.database_name, create.to_table_id.table_name};
+        if (!to_table.table.empty())
+        {
+            /// TO target_table (for materialized views)
+            if (to_table.database.empty())
+                to_table.database = data.default_database;
+            data.dependencies.emplace(to_table);
+        }
 
-    return true;
-}
+        QualifiedTableName as_table{create.as_database, create.as_table};
+        if (!as_table.table.empty())
+        {
+            /// AS table_name
+            if (as_table.database.empty())
+                as_table.database = data.default_database;
+            data.dependencies.emplace(as_table);
+        }
+    }
 
-ssize_t DDLMatcherBase::getPositionOfTableNameArgument(const ASTFunction & function)
-{
-    if (function.name == "joinGet" ||
-        function.name == "dictHas" ||
-        function.name == "dictIsIn" ||
-        function.name.starts_with("dictGet"))
-        return 0;
+    /// ASTTableExpression represents a reference to a table in SELECT query.
+    /// DDLDependencyVisitor should handle ASTTableExpression because some CREATE queries can contain SELECT queries after AS
+    /// (for example, CREATE VIEW).
+    void visitTableExpression(const ASTTableExpression & expr, DDLDependencyVisitor::Data & data)
+    {
+        if (!expr.database_and_table_name)
+            return;
```

Comment:
Could you please add a test with queries like `CREATE VIEW ... AS SELECT ... FROM (SELECT ... FROM table)`? (I guess it will just visit another `ASTTableExpression`, and it will work fine, but it's good to have more complex tests)

Reply:
Will do

Reply:
I added the test in https://github.com/ClickHouse/ClickHouse/pull/44158